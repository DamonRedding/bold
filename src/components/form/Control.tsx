import * as React from 'react'
import * as classnames from 'classnames'
import { helpersClassnames, UtilProps } from '../../util/Util'
import { Icon } from '../elements/Icon'

export interface ControlProps extends UtilProps {
    className?: string
    icon?: string
    iconRight?: boolean
}


export class Control extends React.Component<ControlProps, any> {

    render() {
        const controlClasses = classnames('control', this.props.className, helpersClassnames(this.props), {
            'has-icon': this.props.icon !== undefined,
            'has-icon-right': this.props.iconRight
        })

        return (
            <div className={controlClasses}>
                {this.props.children}
                {this.props.icon && <span className='icon'><Icon icon={this.props.icon} /></span>}
            </div>
        )
    }

}
