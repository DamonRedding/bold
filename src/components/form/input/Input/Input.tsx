import * as React from 'react'

export interface InputController {
  getInput(): HTMLInputElement
  focus(): any
  blur(): any
}

export interface PublicInputProps {
  className?: string
  checked?: boolean
  disabled?: boolean
  id?: string
  maxLength?: number
  name?: string
  onBlur?: <T>(event?: React.FocusEvent<T>) => void
  onChange?: <T>(event: React.ChangeEvent<T> | any) => void
  onFocus?: <T>(event?: React.FocusEvent<T>) => void
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  placeholder?: string
  value?: any
  provideController?: (controller: InputController) => void
  autoFocus?: boolean
}

export interface InputProps extends PublicInputProps {
  type?: string
}

export class Input extends React.Component<InputProps, any> {
  private input: HTMLInputElement

  componentDidMount() {
    this.props.provideController && this.props.provideController({
      getInput: () => this.input,
      focus: () => this.input.focus(),
      blur: () => this.input.blur(),
    })
  }

  componentWillUnmount() {
    this.props.provideController && this.props.provideController(null)
  }

  render() {
    const { provideController, ...rest } = this.props

    return (
      <input ref={this.ref} {...rest} />
    )
  }

  private ref = (elem: HTMLInputElement) => {
    this.input = elem
  }

}
