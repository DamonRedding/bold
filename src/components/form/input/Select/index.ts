export * from './SelectMenu'
export * from './Select'
