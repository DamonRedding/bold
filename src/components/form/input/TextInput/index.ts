export { TextInputBase, TextInputBaseProps, InputStatus } from './TextInputBase'
export { TextInput, TextInputProps } from './TextInput'
