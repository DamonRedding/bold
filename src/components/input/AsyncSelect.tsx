import * as React from 'react'
import * as Select from 'react-select'
import axios from 'axios'

export interface AsyncSelectProps extends Partial<Select.ReactAsyncSelectProps> {
    url: string
    value?: any
    pageSize?: number
    searchDelay?: number
    customizeParams?: any,
    maxLength?: number
}

export class AsyncSelect extends React.Component<AsyncSelectProps> {
    private typingTimer: number

    static defaultProps: Partial<AsyncSelectProps> = {
        autoload: false,
        pageSize: 10,
        searchDelay: 500,
        valueKey: 'id',
        labelKey: 'nome',
        multi: false,
        cache: false,
        placeholder: '',
        clearValueText: 'Limpar seleção',
        loadingPlaceholder: 'Carregando...',
        searchPromptText: 'Digite para pesquisar',
        noResultsText: 'Nenhum item encontrado.',
        backspaceRemoves: false,
        ignoreAccents: false,
        ignoreCase: true,
        clearable: true,
    }

    render() {
        const { value, onBlur, ...rest } = this.props
        return (
            <Select.Async
                options={[]}
                {...rest}
                inputProps={{ maxLength: this.props.maxLength }}
                value={value && value.toJS ? value.toJS() : value}
                onBlur={this.blur()}
                loadOptions={this.loadOptions}
            />
        )
    }

    private loadOptions = (query, callback) => {
        clearTimeout(this.typingTimer)
        if (this.typingTimer) {
            clearTimeout(this.typingTimer)
        }
        this.typingTimer = setTimeout(this.getPage.bind(this, query, callback), this.props.searchDelay)
    }

    private getPage(query, callback) {
        let params = {
            query: query,
            pageSize: this.props.pageSize
        }

        params = this.props.customizeParams ? this.props.customizeParams(params) : params

        return axios.get(this.props.url, { params: params })
            .then(data => {
                let response = {
                    options: data.data
                }
                callback(null, response)
            }).catch(error => {
                callback(error, null)
            })
    }

    private blur = (): React.EventHandler<React.FocusEvent<{}>> => {
        return () => {
            if (this.props.onBlur) {
                if (this.props.value && this.props.value[this.props.labelKey]) {
                    this.props.onBlur(this.props.value)
                } else {
                    this.props.onBlur(null)
                }
            }
        }
    }
}
