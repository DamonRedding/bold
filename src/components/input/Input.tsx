import * as React from 'react'
import { EventOrValueHandler } from 'redux-form'
import * as classnames from 'classnames'
import { helpersClassnames, extractProps } from '../../util/Util'
import { ChangeEvent, FocusEvent } from 'react'

export interface PublicInputProps {
  className?: string
  checked?: boolean
  disabled?: boolean
  id?: string
  maxLength?: number
  name?: string
  onBlur?: EventOrValueHandler<FocusEvent<any>>
  onChange?: EventOrValueHandler<ChangeEvent<any>>
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  placeholder?: string
  value?: any
}

export interface InputProps extends PublicInputProps {
  type: string
}

const propsToKeep = [
  'checked',
  'disabled',
  'id',
  'maxLength',
  'name',
  'onBlur',
  'onChange',
  'onKeyPress',
  'placeholder',
  'type',
  'value'
]

export class Input extends React.Component<InputProps, any> {
  public input: HTMLInputElement

  private ref = (elem: HTMLInputElement) => {
    this.input = elem
  }

  focus() {
    this.input.focus()
  }

  blur() {
    this.input.blur()
  }

  render() {
    const classes = classnames('input', this.props.className, helpersClassnames(this.props))

    return (
      <input ref={this.ref} {...extractProps(this.props, ...propsToKeep) } className={classes} />
    )
  }

}
