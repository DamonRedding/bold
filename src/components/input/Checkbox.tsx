import * as React from 'react'
import * as classnames from 'classnames'
import { WrappedFieldInputProps } from 'redux-form'
import { Input } from './Input'
import { Icon } from '../elements/Icon'


export interface CheckboxProps extends Partial<WrappedFieldInputProps> {
    checked?: boolean
    disabled?: boolean
    label?: string
}

export class Checkbox extends React.Component<CheckboxProps, any> {

    render() {
        const classes = classnames('checkbox', {
            'disabled': this.props.disabled
        })
        const { label, ...rest } = this.props
        return (
            <label className={classes} >
                <Input {...rest} type='checkbox' />
                <Icon className='check-indicator' icon='checkbox-check' />
                <span className='check-label'>{this.props.label || this.props.children}</span>
            </label>
        )
    }

}
