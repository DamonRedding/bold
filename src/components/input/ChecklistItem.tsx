import * as React from 'react'
import { CheckboxProps, Checkbox } from './Checkbox/Checkbox'
import { List } from 'immutable'

export interface ChecklistItemProps extends CheckboxProps {
    optionValue: any
}

export class ChecklistItem extends React.Component<ChecklistItemProps> {

    getCurrentValues = () => {
        const { value } = this.props

        let previousValues = []

        if (value !== undefined && value !== '') {
            previousValues = value
        }

        const currentValues = Array.isArray(previousValues) ? [...previousValues] : List.isList(previousValues) ? previousValues : [previousValues]

        return currentValues
    }

    handleChange = (event) => {
        const { onChange } = this.props
        let values: any = this.getCurrentValues()

        if (event.target.checked) {
            if (List.isList(values)) {
                values = values.push(this.props.optionValue)
            } else {
                values.push(this.props.optionValue)
            }
        } else {
            if (List.isList(values)) {
                values = values.splice(values.indexOf(this.props.optionValue), 1)
            } else {
                values.splice(values.indexOf(this.props.optionValue), 1)
            }
        }

        return onChange(values as any)
    }

    render() {
        const { optionValue, ...rest } = this.props
        const values = this.getCurrentValues()

        const isChecked = values.indexOf(optionValue) > -1
        return (
            <Checkbox
                {...rest}
                onChange={this.handleChange}
                checked={isChecked}
                onBlur={this.onBlur}
                value={optionValue}
            />
        )
    }

    private onBlur = () => { }
}
