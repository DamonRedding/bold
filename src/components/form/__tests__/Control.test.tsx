import * as React from 'react'
import { shallowRenderAndMatch } from '../../../__tests__/test.defaults'
import { Control } from '../Control'
import { Input } from '../../input/Input'


describe('Control', () => {
    it('deve renderizar corretamente', function () {
        shallowRenderAndMatch(<Control />)
    })
    it('deve renderizar o ícone', function () {
        shallowRenderAndMatch(<Control icon='login' />)
    })
    it('deve renderizar o ícone na direita', function () {
        shallowRenderAndMatch(<Control icon='login' iconRight />)
    })
    it('deve renderizar os filhos', function () {
        shallowRenderAndMatch(<Control><Input type='text' /></Control>)
    })
})
