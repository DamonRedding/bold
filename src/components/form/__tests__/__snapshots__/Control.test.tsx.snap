// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Control deve renderizar corretamente 1`] = `
<div
  className="control"
/>
`;

exports[`Control deve renderizar o ícone 1`] = `
<div
  className="control has-icon"
>
  <span
    className="icon"
  >
    <Icon
      icon="login"
    />
  </span>
</div>
`;

exports[`Control deve renderizar o ícone na direita 1`] = `
<div
  className="control has-icon has-icon-right"
>
  <span
    className="icon"
  >
    <Icon
      icon="login"
    />
  </span>
</div>
`;

exports[`Control deve renderizar os filhos 1`] = `
<div
  className="control"
>
  <Input
    type="text"
  />
</div>
`;
